tracker/server.go

```diff
@@ -0,0 +1,237 @@
+
+func NewGRPCServer(ctx context.Context, lts radio.ListenerTrackerService) *grpc.Server {
+	gs := rpc.NewGrpcServer(ctx)
+	rpc.RegisterListenerTrackerServer(gs, rpc.NewListenerTracker(lts))
+	return gs
+}
+
+func NewServer(ctx context.Context, cfg config.Config) (*Server, error) {
+	const op errors.Op = "tracker.NewServer"
+
+	// setup recorder
+	var recorder = NewRecorder(ctx, cfg)
+
+	// setup periodic task to update the manager of our listener count
+	go PeriodicallyUpdateListeners(ctx, cfg.Manager, recorder, UpdateListenersTickrate)
+	// setup periodic task to keep recorder state in sync with icecast
+	go PeriodicallySyncListeners(ctx, cfg, recorder, SyncListenersTickrate)
+
+	srv := new(Server)
+	r := website.NewRouter()
+
+	r.Use(
+		hlog.NewHandler(*zerolog.Ctx(ctx)),
+		hlog.RemoteAddrHandler("ip"),
+		hlog.UserAgentHandler("user_agent"),
+		hlog.RequestIDHandler("req_id", "Request-Id"),
+		hlog.URLHandler("url"),
+		hlog.MethodHandler("method"),
+		hlog.ProtoHandler("protocol"),
+		hlog.AccessHandler(zerologLoggerFunc),
+	)
+	r.Post("/listener_joined", ListenerAdd(ctx, recorder))
+	r.Post("/listener_left", ListenerRemove(ctx, recorder))
+
+	var err error
+
+	// setup the GRPC server that the rest will be poking
+	srv.rpcSrv = NewGRPCServer(ctx, recorder)
+	// and a listener for the GRPC server
+	srv.rpcLn, err = net.Listen("tcp", cfg.Conf().Tracker.RPCAddr)
+	if err != nil {
+		return nil, errors.E(op, err)
+	}
+
+	srv.httpSrv = &http.Server{
+		Addr:         cfg.Conf().Tracker.ListenAddr,
+		Handler:      r,
+		ReadTimeout:  time.Second * 10,
+		WriteTimeout: time.Second * 5,
+		BaseContext:  func(l net.Listener) context.Context { return ctx },
+	}
+	srv.httpLn, err = net.Listen("tcp", cfg.Conf().Tracker.ListenAddr)
+	if err != nil {
+		return nil, errors.E(op, err)
+	}
+
+	return srv, nil
+}
+
+func (s *Server) Close() error {
+	s.rpcSrv.Stop()
+	return s.httpSrv.Close()
+}
+
+func (s *Server) Start(ctx context.Context) error {
+	logger := zerolog.Ctx(ctx)
+	errCh := make(chan error, 2)
+	go func() {
+		s.mu.Lock()
+		ln := s.rpcLn
+		s.mu.Unlock()
+		logger.Info().Str("address", ln.Addr().String()).Msg("tracker started listening [grpc]")
+		errCh <- s.rpcSrv.Serve(ln)
+	}()
+	go func() {
+		s.mu.Lock()
+		ln := s.httpLn
+		s.mu.Unlock()
+		logger.Info().Str("address", ln.Addr().String()).Msg("tracker started listening [http]")
+		errCh <- s.httpSrv.Serve(ln)
+	}()
+
+	select {
+	case <-ctx.Done():
+		return s.Close()
+	case err := <-errCh:
+		return err
+	}
+
+}
+
+func (srv *Server) handleResume(ctx context.Context, cfg config.Config) error {
+	parent, err := graceful.FD2Unix(3)
+	if err != nil {
+		return err
+	}
+	defer parent.Close()
+
+	return srv.readSelf(ctx, cfg, parent)
+}
+
+func (srv *Server) handleRestart(ctx context.Context, cfg config.Config) error {
+	dst, err := graceful.StartChild()
+	if err != nil {
+		return err
+	}
+	defer dst.Close()
+
+	return srv.writeSelf(dst)
+}
+
+type httpWire string
+type rpcWire string
```

Comment:
would just make these structs with one field each, makes it easier to add to them later if required.